### DISCUSSION TITLE
add max_devices and name: option to Energy panel

### DISCUSSION THREAD

**Discussion by Mariusthvdb** [home-assistant/frontend#19983] [Dashboard]
like we now (2024.3) will have for the user dashboard cards set in yaml by:

```
type: energy-devices-detail-graph
max_devices: 10
```

<img width="504" alt="Scherm­afbeelding 2024-03-04 om 22 37 32" src="https://github.com/home-assistant/frontend/assets/33354141/b5d9294f-e2c4-422c-9fdf-bfef860731c4">

having many (over 20/30) individual energy entities configured, the individual panel is so crammed it barely is usable, as the card first shows all of those entities, many with longer names, and next graphing them in a single bar....

Allowing the max_devices in the Energy panel would mean a huge difference.

**Comment by karwosts** [comment]
I'll think about this.

It will be more difficult than adding options to regular cards, since we can't just add any old unstructured options to a yaml file like we can do for a regular dashboard. We will have to have changes in both frontend, and changes in core to have an API that we can communicate the setting to the backend, a place for the backend to store the setting, and a way for the backend to return the setting to the frontend when loading the panel. 

I feel like I remember seeing somewhere some maintainers saying that the energy panel was not really intended to be configurable, so we might get pushback if we start trying to add all kinds of extra configuration options like for regular cards. But maybe one tiny setting can be ok.

**Reply by Mariusthvdb** [reply]
agreed, especially since this is truly about core and default purposefulness .

we need to add all devices to have a correct count of the usage, and add even the tiniest of energy users. However for the detailed view, we need only the top contenders, and all of the tiny ones only clutter the view with (untouchable) lines, because they are way too close to each other.

anyways, no need to explain, its clear as it is.
Hope you can make it.
thanks for considering

**Reply by karwosts** [reply]
> we need to add all devices to have a correct count of the usage, and add even the tiniest of energy users

I'm curious about this statement. What do you mean the "correct count of the usage"? I don't believe the sum of these values is shown anywhere, is it? 

I was thinking if we added the max_devices option it would apply to both device graphs, since one could easily call them "clutter" on the original device graph as well. But then I'm thinking I'm not really sure the purpose of adding these small consumers in the first place, as then they will mostly just never be shown anywhere, kind of defeating the point.

**Reply by Mariusthvdb** [reply]
There is no clutter in the existing monitor individual devices, because they are individual. Admittedly a lot of them but no clutter. ( well except for the labels)

As a matter of fact the declining graphs show a very nice overview 

the detailed graph is just about the opposite of ‘clear’

if you’d ask me, the optional setting to limit the number would only be needed for the detailed graph.

resulting in the perfect combination of an overview of all, and detailed for the largest users. That detail then is truly visible

**Comment by Mariusthvdb** [comment]
just realized we could also clear a lot of the clutter, if we could set a name: for the entities, and not have it show the full entity name as set in the device.

ive considered changing those names themselves, but that would not be optimal, as in other places we need to see what the full name.

However, in the energy panel, we have only energy entities, and could bring down the name to a much shorter identifier

eg from 

<img width="1161" alt="Scherm­afbeelding 2024-03-07 om 12 03 35" src="https://github.com/home-assistant/frontend/assets/33354141/b5b63203-a61f-4575-ad80-cc4462f70307">

'Patchboard Zolder: totaal' which is required to make it stand out against 'Patchboard Zolder: actueel' and some other in the system, to a mere 'Patchboard' for the graphs in the energy panel.

this would take down the number of characters for all of those entities significantly, and allow for more spacing for the graphs themselves. Also for the individual graphs btw:

<img width="1153" alt="Scherm­afbeelding 2024-03-07 om 12 07 01" src="https://github.com/home-assistant/frontend/assets/33354141/8d310432-954d-4397-9e4a-cee5d8a1efd8">

**Reply by trvrnrth** [reply]
Some other options I had wondered about:
- Automatically strip common prefixes (`energy`, `Total daily energy` etc.). I guess the problem with this approach is that it could be difficult to compile a suitably complete list with translations etc.
- Instead of the entity name, show the parent device name. This would need to apply only for the case that a single entity from that device is added (eg. I have two energy sensors from the same boiler device).

**Comment by ildar170975** [comment]
Instead of adding some limit - why not just enable hide/show a legend?

**Reply by Mariusthvdb** [reply]
yes, mentioned that before too, would be a very fine improvement indeed
not having to scroll pst all of those entities would be very helpful, especially since the popups show what the entity is anyways

**Comment by OwlSaver** [comment]
Some other options that do not require a setting are:

- Show the top 10 and group the rest as 'All Other'
- Show all with a value greater than the mean and group the rest as 'All Other'

The values 10 or mean could be different but I think could be a fixed value so no setting is needed. I think it is always important to include 'All Other'.

**Reply by erkr** [reply]
If the total usage is from i.e a DSMR meter, the individual measurements will typically not count up to the total (not everything is measured).
So just show the biggest contributors proportional to the DSMR total is perfectly fine then.  And you don’t need to have the small ones in the card to have to correct total either (as mentioned above)

**Comment by LeiChat** [comment]
+1 on the request to add custom entity names for the legend, please.

Similar to the optional "name" attribute on the statistics-graph card.

https://community-assets.home-assistant.io/original/4X/a/7/7/a77bdf8786936770f86cd92d665f8b20fea182fe.png

**Comment by chrismaki** [comment]
> having many (over 20/30) individual energy entities configured, the individual panel is so crammed it barely is usable, as the card first shows all of those entities, many with longer names, and next graphing them in a single bar....
> 
> Allowing the max_devices in the Energy panel would mean a huge difference.

I'd also like to see a way to either exclude/include specific devices or limit the amount displayed to the X-highest consumers.

<img width="1147" alt="image" src="https://github.com/home-assistant/frontend/assets/17991816/30589ec6-353d-4383-8873-94759c3a96e7">

**Comment by erkr** [comment]
This thread started with the request to limit the number of devices in the Detailed devices energy graph. Agreed, good solution for the default card in the standard energy dashboard.
But how to re-use this nice card with different selections of available energy sensors in our own dashboards? 
We would need to configure lists of entities in the card UI editor for that. As this would be UI configuration only, can that by done without impact on the energy  dashboard / yaml configuration?!
Finally population the card using auto-entities is my ultimate goal 😊


Background: 
I grouped all my entity sensors into lights, media, appliances, climate, etc… and use them in the standard energy dashboard. 
now I want to make cards per group, to display the energy distributions in those groups. E.g in the Appliance group, there are the fridges, washing machine, dryer etc

**Comment by erkr** [comment]
This is an example how I used apex-charts to show energy usage of the main groups, or more detailed sub groups as appliances or climate.
‘delta’ in gray is the difference between measured and the total usage of my P1 meter:
![image](https://github.com/home-assistant/frontend/assets/4363599/8ba8e6af-54c5-45c8-aebb-67a26b509f79)
I would love to use the detailed device energy card for that with it’s selectable time span

**Reply by lukaszswitaj** [reply]
@erkr Waiting for a native solution... Can you share the recipe to calculate Delta? Did you use 'data-generator' in Apex Chart Card or template?

**Reply by erkr** [reply]
@lukaszswitaj 
Sorry  I don’t know how to calculate i.e the total of the measures series in apex. 
So I used both a template sensor and an integration sensor for that. 
The template sensor is triggered for each P1 monitor update (actual total power) and subtracts the measured powers. 
The Riemann integration sensor uses this template sensor as an input.

**Comment by Mariusthvdb** [comment]
Please keep this discussion ‘on’ topic and not divert from the actual feature request.

**Comment by erkr** [comment]
> Please keep this discussion ‘on’ topic and not divert from the actual feature request.

That is my intention. I'm in favour of a generic solution, where the cards entities can be specified. So similar to the statistics-graph and many other cards:
``` 
entities:
  - entity: sensor.x
    name: labelX 
  ...
``` 
That solves both the request to specify a short name, and limiting the number of entries per card.

**Reply by MarkusReuter** [reply]
If there would be the possibility to define certain entities, we could use the energy-devices-detail-graph to show only the specified devices. If entities are not specified, every entity is shown as usual.

In my case I have about 80 entities in my energy dashboard. In my special heat-pump dashboard I want to use the graph only for one device, my heat-pump. As long as the heat-pump has the highest consumption, i can use

type: energy-devices-detail-graph
max_devices: 1

But that is not working when other devices have higher consumptions and move to the top of the list. Then ic can't see the data of my heat-pump. Please consider adding an entities parameter to energy-devices-detail-graph.

**Reply by karwosts** [reply]
Couldn't you get the same with just a statistics-graph-card of your heat pump?

**Reply by LeiChat** [reply]
> I want to use the graph only for one device, my heat-pump.

This card is designed for comparison though. If you just want a graph to chart consumption for a single entity/device then you might be better off using one of the various other graph cards.

**Reply by MarkusReuter** [reply]
found out today, that statistics-graph can use the dateselection from the energy board (energy_date_selection: true). Therefore my problem is solved. Thanks for the advice.

**Comment by karwosts** [comment]
Custom names for individual energy devices merged for 2024.5 ([#20033](https://github.com/home-assistant/frontend/pull/20033))

Hopefully this will mitigate some of the large legend issues.
